### PR TITLE
restruct config initialization

### PR DESCRIPTION
**Pull request by dai1975**
# change config initialization structure to use logger in chain and/or prover init

before: 

```
 PersistentPreRunE
 +-- InitConfig
 |   +-- UnmarshalJSON
 |   |   +-- json.Unmarshal
 |   |   +-- config.Chains[].Init
 |   |   +-- config.chains = config.Chains[].Build
 |   +-- InitChains
 |      +-- config.chains[].Init
 +-- initLogger
```

after:
```
 PersistentPreRunE
 +-- UnmarshalConfig
 |   +-- json.Unmarshal
 +-- initLogger
 +-- InitConfig
 |   +-- config.Chains[].Init
 |   +-- config.chains = config.Chains[].Build
 |   +-- initChains
 |      +-- config.chains[].Init
```

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -128,7 +128,7 @@ func (c *Config) ChainsFromPath(path string) (map[string]*core.ProvableChain, st
 }
 
 // Called to initialize the relayer.Chain types on Config
-func InitChains(ctx *Context, homePath string, debug bool) error {
```
Comment (dai1975) [inline config/config.go:131]: This function called by only this file.

File: config/config.go [hunk 2]
```diff
@@ -143,6 +143,46 @@ func InitChains(ctx *Context, homePath string, debug bool) error {
...
+		}
+	} else {
+		defConfig := defaultConfig(cfgPath)
+		*c = defConfig
+	}
+	c.ConfigPath = cfgPath
+	return nil
+}
+
+func (ctx *Context) InitConfig(homePath string, debug bool) error {
```
Comment (dai1975) [inline config/config.go:165]: Old function signature `(c *Config) InitConfig(ctx *Context, ...)` is a bit confusing that c and ctx.Config should be same pointer or not.

File: config/config.go [hunk 3]
```diff
@@ -143,6 +143,46 @@ func InitChains(ctx *Context, homePath string, debug bool) error {
...
+
+	// ensure config has []*relayer.Chain used for all chain operations
+	if err := initChains(ctx, homePath, debug); err != nil {
+		return err
+	}
+	ctx.Config.InitCoreConfig()
+	return nil
+}
+
+/*
```
Comment (siburu) [inline config/config.go:185]: Please delete this code that is commented out.

File: config/marshaler.go [hunk 4]
```diff
@@ -2,27 +2,12 @@ package config
```
Comment (siburu) [inline config/marshaler.go:2]: Please delete this file and use `json.Marshal` and `json.Unmarshal` directly.

File: provers/mock/prover.go [hunk 5]
```diff
@@ -27,6 +28,9 @@ func NewProver(chain core.Chain, config ProverConfig) *Prover {
 }
 
 func (pr *Prover) Init(homePath string, timeout time.Duration, codec codec.ProtoCodecMarshaler, debug bool) error {
+	if logger := log.GetLogger(); logger != nil {
```
Comment (siburu) [inline provers/mock/prover.go:31]: I think we should not check `logger != nil`.
If it is `nil`, the program panics, but it's a desirable result.

**Review by siburu — COMMENTED** [review]
@dai1975 Great work overall, thank you! I've left some comments in the code. Could you please review them and make the necessary changes?

File: provers/mock/prover.go (hunk 5 above)
Comment (dai1975) [reply provers/mock/prover.go]: fixed

File: config/marshaler.go (hunk 4 above)
Comment (dai1975) [reply config/marshaler.go]: deleted.

**Review by siburu — APPROVED** [review]
@dai1975 LGTM. Thank you!!

**Pull request merged**